### PR TITLE
Delete all shifts for a schedule if it's updated to get rid of old va…

### PR DESCRIPTION
**Pull request by elainearbaugh**
…lues

### CODE REVIEW

**Comment by elainearbaugh** [comment]
trigger build

**Comment by elainearbaugh** [comment]
@mahendra tested on stage and updating schedules is working correctly now. I wasn't sure where to fire off the celery task--get_success_url() seems wrong, but it's the only way I could think of to start the task after everything is updated...

File: cabot/cabotapp/tasks.py [hunk 1]
```diff
@@ -83,6 +83,14 @@ def update_shifts():
 
 
 @task(ignore_result=True)
+def reset_shifts(schedule_id):
+    from .models import Schedule, update_shifts as _update_shifts, delete_shifts
+    schedule = Schedule.objects.get(id=schedule_id)
+    delete_shifts(schedule)
```
Comment (mahendra) [inline cabot/cabotapp/tasks.py]: Shouldn't these two operations be done within a single db transaction?

File: cabot/cabotapp/tasks.py [hunk 2]
```diff
@@ -83,6 +83,14 @@ def update_shifts():
...
+    _update_shifts(schedule)
```
Comment (mahendra) [inline cabot/cabotapp/tasks.py]: can we rename `_update_shifts` to `update_shifts` ?

File: cabot/cabotapp/tasks.py [hunk 3]
```diff
@@ -83,6 +83,14 @@ def update_shifts():
 
 
 @task(ignore_result=True)
+def reset_shifts(schedule_id):
+    from .models import Schedule, update_shifts as _update_shifts, delete_shifts
```
Comment (mahendra) [inline cabot/cabotapp/tasks.py]: Let us not do this please. Please move imports to the beginning of the file

File: cabot/cabotapp/views.py [hunk 4]
```diff
@@ -836,6 +836,7 @@ class ScheduleUpdateView(LoginRequiredMixin, UpdateView):
     context_object_name = 'schedules'
 
     def get_success_url(self):
+        reset_shifts.delay(self.object.id)
```
Comment (mahendra) [inline cabot/cabotapp/views.py]: Two comments

1. This is not the right place to call it. Can you please update the PR description with more context? Will help me understand the problem exactly (I am sorry. I know you explained it, but I was so tired yesterday that I forgot. old age!!)
1. `delay` is a deprecated celery api. We should be using `apply_async`

Comment (mahendra) [inline cabot/cabotapp/views.py]: ok. figured it out by reading through the code.

Correct me if I am wrong. Your need is to fire off the task when the `Schedule` model is saved. There are multiple ways to do this.
1. Override the `save` method. Acceptable practice in django, but not clean. Causes business logic to creep into models.
1. Use django signals - https://docs.djangoproject.com/en/1.10/topics/signals/ - Django signals allow you to hook into model actions. Allows for good clean separation!

File: cabot/cabotapp/tasks.py (hunk 2 above)
Comment (elainearbaugh) [inline cabot/cabotapp/tasks.py]: update_shifts is the name of another task, but it has no args so sure?

File: cabot/cabotapp/views.py (hunk 4 above)
Comment (elainearbaugh) [inline cabot/cabotapp/views.py]: Yep that's what I want, thanks for the suggestions!

**Comment by elainearbaugh** [comment]
@mahendra made the changes. going to test on stage now

**Pull request merged**